# 简介
部分功能模块并不支持协程，它们只提供了接口回调风格的API，如果我们直接在协程代码中调用这些接口，就会增加嵌套层级、降低代码的可维护性。

Kotlin提供了 `suspendCoroutine()` 等工具，用于将接口回调风格的API转换为挂起函数，以便在协程代码中调用，减少嵌套层级并统一编码风格。

# 接口回调风格API
下文示例代码展示了一个模拟网络请求API的代码风格：

"TestWorkWithThread.kt":

```kotlin
// 网络请求回调接口
interface NetCallback {

    // 请求成功
    fun onSuccess(data: String)

    // 请求失败
    fun onFailure(message: String)
}

/**
 * 模拟网络请求。
 *
 * @param[result] 控制请求结果。
 * @param[callback] 结果回调。
 */
fun request(result: Boolean, callback: NetCallback) {
    // 开启新线程进行网络请求
    thread {
        // 线程休眠，模拟网络延时。
        Thread.sleep(3000L)
        // 休眠结束，调用回调方法反馈执行结果。
        if (result) {
            callback.onSuccess("200 - OK.")
        } else {
            callback.onFailure("502 - Bad Gateway.")
        }
    }
}
```

NetCallback接口定义了两个方法，其中 `onSuccess()` 方法在请求成功时被触发，参数为服务端返回的数据； `onFailure()` 方法则在请求失败时被触发，参数为错误信息。

`request()` 方法用于模拟网络请求，其中开启了一个线程，休眠3秒以模拟网络耗时，休眠结束后根据参数 `result` 调用NetCallback实例的对应方法，向调用者反馈请求结果。

接口定义完毕后，调用者需要传入NetCallback实现类的实例，并在结果返回时执行对应的操作，例如更新界面等；此处我们只是简单的将回调结果输出到控制台上。

"TestWorkWithThread.kt":

```kotlin
println("Mock request start. Time:[${getTime()}]")
request(true, object : NetCallback {
    override fun onSuccess(data: String) {
        println("OnSuccess. Time:[${getTime()}] Data:[$data]")
    }

    override fun onFailure(message: String) {
        println("OnFailure. Time:[${getTime()}] Info:[$message]")
    }
})
```

此时运行示例程序，并查看控制台输出信息：

```text
Mock request start. Time:[23:18:09.453]
OnSuccess. Time:[23:18:12.481] Data:[200 - OK.]
```

根据上述输出内容可知：

`request()` 方法被调用3秒后，NetCallback实例的 `onSuccess()` 方法被触发，输出了请求结果。

# "suspendCoroutine()"方法
## 基本应用
`suspendCoroutine()` 方法是一种允许开发者控制协程任务暂停与恢复的工具，我们可以用它将前文的回调风格接口 `request()` 转换为挂起函数。

"TestWorkWithThread.kt":

```kotlin
// 模拟网络请求（挂起函数实现）。
private suspend fun requestSuspend(result: Boolean): String {
    return suspendCoroutine {
        request(result, object : NetCallback {
            override fun onSuccess(data: String) {
                // 请求成功，解除挂起状态并返回数据。
                it.resume(data)
            }

            override fun onFailure(message: String) {
                // 请求失败，抛出异常。
                val exception = Exception(message)
                it.resumeWithException(exception)
            }
        })
    }
}
```

当协程任务运行至 `suspendCoroutine()` 方法时，进入暂停状态，随后调度器开始执行模拟网络请求。当回调方法 `onSuccess()` 通知请求成功事件时，我们需要调用 `it.resume()` 方法通知协程任务从暂停状态恢复，并传入参数来传递回调方法的请求结果。 `suspendCoroutine()` 方法的返回值是一个泛型类型，与 `it.resume()` 方法的参数类型保持一致，此处请求结果 `data` 的数据类型为String，因此 `suspendCoroutine()` 方法的返回值类型也是String。

接下来，我们就可以在协程任务中调用 `requestSuspend()` 挂起函数，以协程的方式使用接口回调API。

"TestWorkWithThread.kt":

```kotlin
CoroutineScope(Dispatchers.IO).launch {
    println("Mock request start. Time:[${getTime()}]")
    // 声明变量以便接收请求成功的结果
    val data = requestSuspend(true)
    println("Request success. Time:[${getTime()}] Data:[$data]")
}
```

此时运行示例程序，并查看控制台输出信息：

```text
Mock request start. Time:[21:18:09.453]
Request success. Time:[21:18:12.481] Data:[200 - OK.]
```

根据上述输出内容可知：

协程任务运行至 `suspendCoroutine()` 方法时，进入暂停状态，调度器开始执行模拟网络请求；3秒后，回调接口触发了 `onSuccess()` 回调方法及其中的 `it.resume()` 方法，调用者接收到请求结果并从暂停状态恢复。

## 反馈失败结果
NetCallback接口有“请求成功”和“请求失败”两种情况，由于挂起函数只能有单个返回值，已经被用来反馈成功状态，我们需要通过异常机制反馈失败状态给调用者。

"TestWorkWithThread.kt":

```kotlin
CoroutineScope(Dispatchers.IO).launch {
    try {
        println("Mock request start. Time:[${getTime()}]")
        // 模拟请求失败的情况
        val data = requestSuspend(false)
        println("Request success. Time:[${getTime()}] Data:[$data]")
    } catch (e: Exception) {
        // 捕获异常以获取失败详情
        println("Request failure. Info:[${e.message}]")
    }
}
```

此时运行示例程序，并查看控制台输出信息：

```text
Mock request start. Time:[21:22:02.464]
Request failure. Time:[21:22:05.580] Info:[502 - Bad Gateway.]
```

根据上述输出内容可知：

模拟网络请求开始3秒后，触发了 `onFailure()` 回调方法及其中的 `it.resumeWithException()` 方法，向调用者抛出异常信息；随后调用 `requestSuspend()` 方法的协程任务捕获了异常信息，并且能够通过Exception实例获取异常详情。

## 注意事项
下文列表列出了 `suspendCoroutine()` 方法的注意事项：

- `it.resume()` 语句只能被调用一次，否则会导致异常： `IllegalStateException: Already resumed` 。
- 如果 `it.resume()` 语句之后还有其他语句，当 `it.resume()` 语句被调用后，剩余的语句仍然会被执行。

# "runBlocking()"方法
`runBlocking()` 方法的功能与CoroutineScope的 `launch()` 和 `async()` 方法相同，都可以提供协程环境并开启协程任务，但 `launch()` 和 `async()` 方法属于CoroutineScope实例，且不会阻塞调用者线程； `runBlocking()` 方法则可以被直接调用，不需要依赖CoroutineScope实例，并且会阻塞调用者线程，直到协程任务运行完毕。

该方法的主要用途是将协程API转换为接口回调API，以便在不支持协程的环境中使用，或者在 `main()` 方法中测试协程API。

我们可以使用 `runBlocking()` 方法再将前文示例中模拟网络请求的挂起函数转换为接口回调形式：

"TestWorkWithThread.kt":

```kotlin
// 模拟网络请求（协程转接口回调）。
fun requestCallback(result: Boolean, callback: NetCallback) {
    // 在新线程进行网络请求
    thread {
        // 开启协程任务并阻塞当前线程
        runBlocking {
            try {
                val data: String = requestSuspend(result)
                // 获取到结果后，通过回调方法通知调用者。
                callback.onSuccess(data)
            } catch (e: Exception) {
                callback.onFailure(e.message ?: "")
            }
        }
    }
}

// 测试方法
fun main() {
    println("Mock request start. Time:[${getTime()}]")
    requestCallback(true, object : NetCallback {
        override fun onSuccess(data: String) {
            println("OnSuccess. Time:[${getTime()}] Data:[$data]")
        }

        override fun onFailure(message: String) {
            println("OnFailure. Time:[${getTime()}] Info:[$message]")
        }
    })
}
```

此时运行示例程序，并查看控制台输出信息：

```text
Mock request start. Time:[14:36:08.621]
OnSuccess. Time:[14:36:11.672] Data:[200 - OK.]
```
